src/dotty/tools/dotc/transform/ElimErasedValueType.scala

```diff
@@ -0,0 +1,97 @@
+package dotty.tools.dotc
+package transform
+
+import ast.{Trees, tpd}
+import core._, core.Decorators._
+import TreeTransforms._, Phases.Phase
+import Types._, Contexts._, Constants._, Names._, NameOps._, Flags._, DenotTransformers._
+import SymDenotations._, Symbols._, StdNames._, Annotations._, Trees._, Scopes._, Denotations._
+import TypeErasure.ErasedValueType, ValueClasses._
+
+class ElimErasedValueType extends MiniPhaseTransform with DenotTransformer {
+
+  import tpd._
+
+  override def phaseName: String = "elimerasedvaluetype"
+
+  override def runsAfter: Set[Class[_ <: Phase]] = Set(classOf[Erasure])
+
+  override def transform(ref: SingleDenotation)(implicit ctx: Context): SingleDenotation = {
+    val info = ref match {
+      case ref: ClassDenotation if ref is ModuleClass =>
+        ref.linkedClass match {
+          case origClass: ClassSymbol if isDerivedValueClass(origClass) =>
+            val cinfo = ref.classInfo
+            val decls1 = cinfo.decls.cloneScope
+            ctx.atPhase(this.next) { implicit ctx =>
+              // Remove synthetic cast methods introduced by ExtensionMethods,
+              // they are no longer needed after this phase.
+              decls1.unlink(cinfo.decl(nme.UNDERLYING2EVT).symbol)
+              decls1.unlink(cinfo.decl(nme.EVT2UNDERLYING).symbol)
+            }
+            cinfo.derivedClassInfo(decls = decls1)
+          case _ =>
+            ref.info
+        }
+      case _ =>
+        ref.info
+    }
+    val info1 = elimErasedValueType(info)
+    if (info1 eq ref.info) ref
+    else ref match {
+      case ref: SymDenotation => ref.copySymDenotation(info = info1)
+      case _ => ref.derivedSingleDenotation(ref.symbol, info1)
+    }
+  }
+
+  override def checkPostCondition(tree: tpd.Tree)(implicit ctx: Context) = {
+    val tp = tree.tpe
+    val erased = elimErasedValueType(tp)
+    assert(tp eq erased,
+      i"The type $tp of tree $tree contains ErasedValueType, it should have been erased to $erased by ElimErasedValueType.")
+    val name = tree.symbol.name
+    assert(name != nme.UNDERLYING2EVT && name != nme.EVT2UNDERLYING,
+      i"The symbol '${tree.symbol}' of tree $tree should not exist after phase ElimErasedValueType.")
+  }
+
+  // FIXME: Is using a TypeMap here the best solution?
+  def elimErasedValueType(tpe: Type)(implicit ctx: Context) = new TypeMap() {
+    def apply(tp: Type) = tp match {
+      case ErasedValueType(_, underlying) =>
+        underlying
+      case _ =>
+        mapOver(tp)
+    }
+  }.apply(tpe)
+
+  def transformTypeOfTree(tree: Tree)(implicit ctx: Context): Tree =
+    tree.withType(elimErasedValueType(tree.tpe))
```

Comment:
you shound't assign types by hand. You should fix the tree so that it has correct type on its own.